---
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    toc_depth: 3
    fig_width: 5
vignette: >
  %\VignetteIndexEntry{Bioc2020RCWL-1}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding[utf8]{inputenc}
---
# Connecting Bioconductor to other bioinformatics tools using `Rcwl`

## Instructor(s) name(s) and contact information

* [Qian Liu](https://github.com/liubuntu) (Qian.Liu@roswellpark.org)
* [Qiang Hu](https://github.com/hubentu) (Qiang.Hu@roswellpark.org)

## Workshop Description

This workshop introduces the Bioconductor toolchain for usage and
development of reproducible bioinformatics pipelines using packages of
Rcwl and RcwlPipelines. The Common Workflow Language (CWL) is an open
standard for development of data analysis workflows that is portable
and scalable across different tools and working environments. Rcwl
provides a simple way to wrap command line tools and build CWL data
analysis pipelines programmatically within R. It increases the ease of
use, development, and maintenance of CWL pipelines, and furthermore
offers higher performance by intuitively supporting parallel work with
high performance computing (HPC). More than a hundred of pre-built and
tested CWL tool/pipeline recipes are included in RcwlPipelines. These
tools and pipelines are highly modularized for easy customization of
complex bioinformatics analysis. Here we included a scRNAseq
preprocessing pipeline which uses `STARsolo` for alignment and
quantification, and `DropletUtils` for filtering raw gene-barcode
matrix and removing empty droplets. This pipeline demonstrates the
typical use case of our packages. More details for usage and examples
are available on Rcwl website: https://hubentu.github.io/Rcwl/.


## Pre-requisites

- Basic knowledge of using R and Bioconductor packages
- Basic familiarity with running command-line tools
- No prior experience with CWL is necessary!

## Workshop Participation

Participants will be able to try out all of the functionality
described. Active user participation throughout the event is highly
encouraged including but not limited to lecture material, hands-on
sections and final discussion.

Some basic idea about how CWL works.
* https://www.commonwl.org/user_guide/

## _R_ / _Bioconductor_ packages used

* [`Rcwl`](https://bioconductor.org/packages/Rcwl/)
* [`RcwlPipelines`](https://bioconductor.org/packages/RcwlPipelines/)
<!-- * [`BiocParallel`](https://bioconductor.org/packages/BiocParallel/) -->
<!-- *[`shiny`](https://CRAN.R-project.org/package=shiny) -->

System dependencies
* [cwltool](https://github.com/common-workflow-language/cwltool)
* [docker](https://docs.docker.com/get-docker/)

## Time outline

| Activity                                                      | Time |
|---------------------------------------------------------------|------|
| Overview of bioinformatics pipelines and CWL (ppt)            | 10 min |
| Use core functions to update, search and load tools/pipelines | 5 min  |
| Use single cell indexing tool `STARindex`                     | 5 min |
| Use single cell alignment tool `STARsolo`                     | 5 min  |
| Use single cell filtering tool `DropletUtils`                 | 5 min  |
| Use single cell preprocessing pipeline `STARsoloDropletUtils` | 5 min  |
| Wrap command line tools using `Rcwl`                          | 0m (developers only) |
| Customize your pipelines using `Rcwl`                         | 0m (developers only) |

## Workshop goals and objectives

### Learning goals

* Basic knowledge of Common Workflow Language (CWL)
* Knowledge of R/Bioconductor interface of CWL
* Usage of the pre-built bioinformatics tools and pipelines in _R_
* Understand how to wrap command line tools with `Rcwl` (developers only)
* Understand how to build bioinformatics pipelines with `Rcwl` (developers only)

### Learning objectives

* Run the scRNA-seq indexing tool `STARindex` in _R_
* Run the scRNA-seq alignment tool `STARsolo` in _R_
* Run the scRNA-seq QC tool `DropletUtils` in _R_
* Run the scRNA-seq preprocessing pipeline which combines alignment and QC steps in _R_
* Create a basic echo tool using `Rcwl` (developers only)
* ?? Build a basic pipeline ??? using `RCWL` (developers only) 

## Introduction to CWL

Please see details in the slides. 

## single cell RNA sequencing data 

### Data resources

We have prepared all the files needed in this workshop in a GitHub
repository, and now we need to download these files to a local
temporary folder for the demo purpose. We will also use the `path`
below as the file path to save all intermediate files.

```{r}
path <- system.file("testdata", package = "Bioc2020RCWL")
outpath <- "~/outdir"
```

The single cell data resource we are using today is the 1k PBMCs from
a Healthy Donor (v3 chemistry) from 10x genomics, consisting of 1000
Peripheral blood mononuclear cells (PBMCs) extracted from a healthy
donor, where PBMCs are primary cells with relatively small amounts of
RNA (~1pg RNA/cell). 

The source material consists of 6 FASTQ files split into two
sequencing lanes L001 and L002, each with three reads of R1
(barcodes), R2 (cDNA sequences), I1 (illumina lane infold). The Cell
Ranger pipeline requires all three files to perform the demultiplexing
and quantification, but RNA STARsolo does not require the I1 lane file
to perform the analysis. The source files can be found
[here](https://zenodo.org/record/3457880#.XxLjOvhKjvU).

For this tutorial, we will use datasets sub-sampled from the source
files to contain only 15 cells instead of 1000. We have also further
curated the data to only include reads on chromosome 21, so that the
real execuation of our CWL tools/pipelines in _R_ can be done within
1~2 minutes for each step. 
```
subset15_chr21_pbmc_1k_v3_S1_L001_R1_001.fastq.gz
subset15_chr21_pbmc_1k_v3_S1_L001_R2_001.fastq.gz
subset15_chr21_pbmc_1k_v3_S1_L002_R1_001.fastq.gz
subset15_chr21_pbmc_1k_v3_S1_L002_R2_001.fastq.gz
```

For the mapping, we need a “whitelist” of known cell barcodes. Here we
used the 15 barcodes that we already subsetted. The original 737,000
barcodes can be freely extracted from the [Cell Ranger
pipeline](https://kb.10xgenomics.com/hc/en-us/articles/115004506263-What-is-a-barcode-whitelist)).

``` 
subset15_demo_barcode.txt 
``` 

The barcodes in the R1 FASTQ data are checked against these known cell
barcodes in order to assign a specific read to a specific known
cell. The barcodes are designed in such a manner that there is
virtually no chance that they will align to a place in the reference
genome. 

In this tutorial we will be using hg19 (GRCh37) version of the
human genome, and will therefore also need to use a hg19 GTF fileto
annotate our reads (on chromosome 21 only).

```
chr21.fa
Homo_sapiens.GRCh37.75.21.gtf
```

10x Genomics has its own processing pipeline, `Cell Ranger` to process
the scRNA-seq outputs it produces, but this process requires much
configuration to run and is significantly slower than other mappers.

### `Rcwl` version of scRNA-seq preprocessing steps

In this tutorial, we are trying to reproduce the [GALAXY
pipeline](https://galaxyproject.github.io/training-material/topics/transcriptomics/tutorials/scrna-preprocessing-tenx/tutorial.html#producing-a-count-matrix-from-fastq)
for scRNA-seq data pre-processing. We will use `STARsolo` to produce a
count matrix from FASTQ, and `DropletUtils` to produce a high quality
count matrix with feature/cell annotation files saved in an R object
of `SingleCellExperiment`. Before these 2 steps, we have also added a
one-time indexing step in our workshop.

#### Loading the tools

With `Rcwl`, we can easily write cwl tools and pipelines
programmatically in _R_ (Will cover more of this topic later for
developers). While we already have the _R_ version of many command
line tools in `RcwlPipelines`, all we need to do is: 1) Search and
load the tools/pipelines, 2) assign values for each of the defined
parameters, 3) execute the command, and 4) get results saved in
user-specified directory, all within _R_.

Here we show the usage of 3 core functions: `cwlUpdate`, `cwlSearch`
and `cwlInstall`. 

##### `cwlUpdate`
The `cwlUpdate` syncs the current Rcwl recipes and return a
`BiocFileCache` object which contains the most updated Rcwl recipes.

```{r}
library(Rcwl)
library(RcwlPipelines)
atls <- cwlUpdate()  ## sync the tools/pipelines
atls
```

the `bfcinfo()` function returns a BiocFileCache tibble containing all
information about each available tool or pipeline. Currently we have
integrated 103 command line tools and 26 pipelines.

```{r}
bfcinfo(atls)
table(bfcinfo(atls)$Type)
```

##### `cwlSearch`

We can use keywords to search for any specific tools/pipelines of
interest, which internally search the keywords against the
BiocFileCache tibble in columns of "rname", "rpath", "fpath",
"Command" and "Containers". Multiple keywords are allowed.

```{r}
tls <- cwlSearch(c("STAR", "index"))
data.frame(tls)
```

The last core function is `cwlInstall` which sources the Rcwl script
for the interested tool/pipeline into the _R_ working environment.

```{r}
cwlInstall(tls$rname[2])  ## "tl_STARindex"
cwlInstall(tls$fpath[2])  ## equivalent to the above. 
STARindex
cwlVersion(STARindex)
inputs(STARindex)
outputs(STARindex)
requirements(STARindex)
```

Now the Rcwl version of `STARindex` is successfully loaded in to _R_
and ready to use. 

#### Indexing

Before the alignment and QC in the aboved mentioned GALAXY pipeline,
we will add an indexing step using the tool `STARindex`. The command
line for indexing looks like this:

```
$ STAR --runMode genomeGenerate --runThreadN 4 --genomeDir STARindex 
--genomeFastaFiles chr21.fa --sjdbGTFfile Homo_sapiens.GRCh37.75.21.gtf
```

Using the `STARindex`, we can equivalently do the same indexing within
_R_ which was internally passed as a cwl script. What we need to do is
assign values for the input parameters, and execute the cwl script
using `runCWL` function.

```{r, eval=FALSE}
inputs(STARindex)
STARindex$genomeFastaFiles <- file.path(path, "chr21.fa")
STARindex$sjdbGTFfile <- file.path(path, "Homo_sapiens.GRCh37.75.21.gtf")
```

```{r, eval=FALSE}
system.time(
    runCWL(cwl = STARindex, outdir = file.path(outpath, "STARindex_output"),
           docker = "singularity", showLog = TRUE)
    ) ## elapsed: 74.795S
dir(file.path(outpath, "STARindex_output"), recursive = TRUE)
```

Now that the output files are in the folder of "STARindex_output"
under the temporary directory. And then these intermediate results are
ready to pass to the next `Rcwl` tool for single cell read alignment. 

#### Alignment

To align single cell reads using `STAR`, it can be done directly in
command line:

```
$ STAR --soloType CB_UMI_Simple --genomeDir STARindex_output/STARindex --soloCBwhitelist subset15_demo_barcode.txt --soloUMIlen 12 --readFilesIn subset15_chr21_pbmc_1k_v3_S1_L001_R2_001.fastq.gz,subset15_chr21_pbmc_1k_v3_S1_L002_R2_001.fastq.gz subset15_chr21_pbmc_1k_v3_S1_L001_R1_001.fastq.gz,subset15_chr21_pbmc_1k_v3_S1_L002_R1_001.fastq.gz --soloUMIfiltering MultiGeneUMI --soloCBmatchWLtype 1MM_multi_pseudocounts --readFilesCommand gzcat --runThreadN 1
```

Now we can follow the previous example to load our Rcwl version of
`STARsolo` and run the command in _R_.

```{r}
cwlInstall("tl_STARsolo")  ## "tl_STARsolo"
STARsolo
inputs(STARsolo)
```
```{r, eval=FALSE}
cdna.fastq <- file.path(path, list.files(path, pattern = "_R2_"))
cb.fastq <- file.path(path, list.files(path, pattern = "_R1_"))
cblist <- file.path(path, "subset15_demo_barcode.txt")
genomeDir <- file.path(path, "STARindex_output/STARindex")

STARsolo$readFilesIn_cdna <- cdna.fastq
STARsolo$readFilesIn_cb <- cb.fastq
STARsolo$whiteList <- cblist
STARsolo$genomeDir <- genomeDir
```

```{r, eval=FALSE}
runCWL(STARsolo, outdir = file.path(outpath, "STARsolo_output"), ,
       docker = "singularity", showLog = TRUE)  ## approximately 2 mins
dir(file.path(outpath, "STARsolo_output"), recursive = TRUE)
```     
Now we have a folder generated in the temporary directory called
"STARsolo_output". The results include ?? which can be passed into
the next tool.

#### `DropletUtils`

To get a high quality count matrix we must apply the `DropletUtils`
tool, which will produce a filtered dataset that is more
representative of the Cell Ranger pipeline.

```{r}       
cwlInstall("tl_DropletUtils")
inputs(DropletUtils)
```
```{r, eval=FALSE}
DropletUtils$dirname <- file.path(outpath, "STARsolo_output/Solo.out")
DropletUtils$lower <- 100
DropletUtils$df <- 5
```

```{r, eval=FALSE}
runCWL(DropletUtils, outdir = file.path(outpath, "dropletUtils_output"), showLog = TRUE)
dir(file.path(outpath, "dropletUtils_output"))
```

Now that we get 2 output files: 

- The pdf file with 2 diagnostic figures: Barcode ranks, and empty droplets. Details for each diagnostic figures please refer to the [DropletUtils vignette](https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html)  
- The `SingleCellExperiment` object which has filtered out unqualified cells.  

```{r, eval=FALSE}
#> Biobase::openPDF(file.path(path, "dropletUtils_output/diagnostics.pdf"))
sce <- readRDS(file.path(outpath, "dropletUtils_output/sce_filtered.rds"))
sce
```
### scRNA-seq preprocessing pipeline

Alternatively and more easily, we can also run the pipeline that was
available in `RcwlPipelines` which has integrated the `STARsolo` and
`DropletUtils` for a streamlined preprocessing analysis within _R_.

```{r, eval=FALSE}
cwlInstall("pl_STARsoloDropletUtils")
inputs(STARsoloDropletUtils)
STARsoloDropletUtils$fastq_cdna <- cdna.fastq
STARsoloDropletUtils$fastq_cb <- cb.fastq
STARsoloDropletUtils$genomeDir <- file.path(outpath, "STARindex_output/STARindex")
STARsoloDropletUtils$whiteList <- cblist
STARsoloDropletUtils$runThreadN <- 2
```

```{r, eval=FALSE}
runCWL(STARsoloDropletUtils, outdir = file.path(outpath, "scpipeline_output"),
       showLog = TRUE)
dir(file.path(outpath, "scpipeline_output"), recursive = TRUE)
outputs(STARsoloDropletUtils)
```

## todo:
- explain the output globs. 
- Explain the cwl files. (config, values)
- test docker. 
