<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • Bioc2020RCWL</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="Bioc2020RCWL">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Bioc2020RCWL</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/Bioc2020RCWL.html">Workshop</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Bioc2020RCWL.html">Bioc2020RCWL (usage)</a>
    </li>
    <li>
      <a href="../articles/Bioc2020RCWL4devel.html">Bioc2020RCWL (developers)</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/liubuntu/Bioc2020Rcwl">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
            
      
      
      <div class="hidden name"><code>Bioc2020RCWL.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html">opts_chunk</a></span>$<span class="fu">set</span>(<span class="kw">message</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<div id="connecting-bioconductor-to-other-bioinformatics-tools-using-rcwl" class="section level1">
<h1 class="hasAnchor">
<a href="#connecting-bioconductor-to-other-bioinformatics-tools-using-rcwl" class="anchor"></a>Connecting Bioconductor to other bioinformatics tools using <code>Rcwl</code>
</h1>
<div id="instructors-names-and-contact-information" class="section level2">
<h2 class="hasAnchor">
<a href="#instructors-names-and-contact-information" class="anchor"></a>Instructor(s) name(s) and contact information</h2>
<ul>
<li>
<a href="https://github.com/liubuntu">Qian Liu</a> (<a href="mailto:Qian.Liu@roswellpark.org" class="email">Qian.Liu@roswellpark.org</a>)</li>
<li>
<a href="https://github.com/hubentu">Qiang Hu</a> (<a href="mailto:Qiang.Hu@roswellpark.org" class="email">Qiang.Hu@roswellpark.org</a>)</li>
</ul>
</div>
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h2>
<p>This workshop introduces the Bioconductor toolchain for usage and development of reproducible bioinformatics pipelines using packages of Rcwl and RcwlPipelines. The Common Workflow Language (CWL) is an open standard for development of data analysis workflows that is portable and scalable across different tools and working environments. Rcwl provides a simple way to wrap command line tools and build CWL data analysis pipelines programmatically within R. It increases the ease of use, development, and maintenance of CWL pipelines, and furthermore offers higher performance by intuitively supporting parallel work with high performance computing (HPC). More than a hundred of pre-built and tested CWL tool/pipeline recipes are included in RcwlPipelines. These tools and pipelines are highly modularized for easy customization of complex bioinformatics analysis. Here we included a scRNAseq preprocessing pipeline which uses <code>STARsolo</code> for alignment and quantification, and <code>DropletUtils</code> for filtering raw gene-barcode matrix and removing empty droplets. This pipeline demonstrates the typical use case of our packages. More details for usage and examples are available on Rcwl website: <a href="https://hubentu.github.io/Rcwl/" class="uri">https://hubentu.github.io/Rcwl/</a>.</p>
</div>
<div id="pre-requisites" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h2>
<ul>
<li>Basic knowledge of using R and Bioconductor packages</li>
<li>Basic familiarity with running command-line tools</li>
<li>No prior experience with CWL is necessary!</li>
</ul>
</div>
<div id="workshop-participation" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h2>
<p>Participants will be able to try out all of the functionality described. Active user participation throughout the event is highly encouraged including but not limited to lecture material, hands-on sections and final discussion.</p>
<p>Some basic idea about how CWL works. * <a href="https://www.commonwl.org/user_guide/" class="uri">https://www.commonwl.org/user_guide/</a></p>
</div>
<div id="r-bioconductor-packages-used" class="section level2">
<h2 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h2>
<ul>
<li><a href="https://bioconductor.org/packages/Rcwl/"><code>Rcwl</code></a></li>
<li>
<a href="https://bioconductor.org/packages/RcwlPipelines/"><code>RcwlPipelines</code></a> <!-- * [`BiocParallel`](https://bioconductor.org/packages/BiocParallel/) --> <!-- *[`shiny`](https://CRAN.R-project.org/package=shiny) -->
</li>
</ul>
<p>System dependencies * <a href="https://github.com/common-workflow-language/cwltool">cwltool</a> * <a href="https://docs.docker.com/get-docker/">docker</a></p>
</div>
<div id="time-outline" class="section level2">
<h2 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h2>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Overview of bioinformatics pipelines and CWL (ppt)</td>
<td>10 min</td>
</tr>
<tr class="even">
<td>Use core functions to update, search and load tools/pipelines</td>
<td>5 min</td>
</tr>
<tr class="odd">
<td>Use single cell indexing tool <code>STARindex</code>
</td>
<td>5 min</td>
</tr>
<tr class="even">
<td>Use single cell alignment tool <code>STARsolo</code>
</td>
<td>5 min</td>
</tr>
<tr class="odd">
<td>Use single cell filtering tool <code>DropletUtils</code>
</td>
<td>5 min</td>
</tr>
<tr class="even">
<td>Use single cell preprocessing pipeline <code>STARsoloDropletUtils</code>
</td>
<td>5 min</td>
</tr>
<tr class="odd">
<td>Wrap command line tools using <code>Rcwl</code>
</td>
<td>0m (developers only)</td>
</tr>
<tr class="even">
<td>Customize your pipelines using <code>Rcwl</code>
</td>
<td>0m (developers only)</td>
</tr>
</tbody>
</table>
</div>
<div id="workshop-goals-and-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h2>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h3>
<ul>
<li>Basic knowledge of Common Workflow Language (CWL)</li>
<li>Knowledge of R/Bioconductor interface of CWL</li>
<li>Usage of the pre-built bioinformatics tools and pipelines in <em>R</em>
</li>
<li>Understand how to wrap command line tools with <code>Rcwl</code> (developers only)</li>
<li>Understand how to build bioinformatics pipelines with <code>Rcwl</code> (developers only)</li>
</ul>
</div>
<div id="learning-objectives" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h3>
<ul>
<li>Run the scRNA-seq indexing tool <code>STARindex</code> in <em>R</em>
</li>
<li>Run the scRNA-seq alignment tool <code>STARsolo</code> in <em>R</em>
</li>
<li>Run the scRNA-seq QC tool <code>DropletUtils</code> in <em>R</em>
</li>
<li>Run the scRNA-seq preprocessing pipeline which combines alignment and QC steps in <em>R</em>
</li>
<li>Create a basic echo tool using <code>Rcwl</code> (developers only)</li>
<li>Build a basic DNAseq pipeline using <code>RCWL</code> (developers only)</li>
</ul>
</div>
</div>
<div id="introduction-to-cwl" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-cwl" class="anchor"></a>Introduction to CWL</h2>
<p>Please see details in the slides.</p>
</div>
<div id="single-cell-rna-sequencing-data" class="section level2">
<h2 class="hasAnchor">
<a href="#single-cell-rna-sequencing-data" class="anchor"></a>single cell RNA sequencing data</h2>
<div id="data-resources" class="section level3">
<h3 class="hasAnchor">
<a href="#data-resources" class="anchor"></a>Data resources</h3>
<p>We have prepared all the files needed in this workshop in a GitHub repository, and now we need to download these files to a local temporary folder for the demo purpose. We will also use the <code>path</code> below as the file path to save all intermediate files.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"testdata"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"Bioc2020RCWL"</span>)
<span class="no">outpath</span> <span class="kw">&lt;-</span> <span class="st">"~/outdir"</span></pre></body></html></div>
<p>The single cell data resource we are using today is the 1k PBMCs from a Healthy Donor (v3 chemistry) from 10x genomics, consisting of 1000 Peripheral blood mononuclear cells (PBMCs) extracted from a healthy donor, where PBMCs are primary cells with relatively small amounts of RNA (~1pg RNA/cell).</p>
<p>The source material consists of 6 FASTQ files split into two sequencing lanes L001 and L002, each with three reads of R1 (barcodes), R2 (cDNA sequences), I1 (illumina lane infold). The Cell Ranger pipeline requires all three files to perform the demultiplexing and quantification, but RNA STARsolo does not require the I1 lane file to perform the analysis. The source files can be found <a href="https://zenodo.org/record/3457880#.XxLjOvhKjvU">here</a>.</p>
<p>For this tutorial, we will use datasets sub-sampled from the source files to contain only 15 cells instead of 1000. We have also further curated the data to only include reads on chromosome 21, so that the real execuation of our CWL tools/pipelines in <em>R</em> can be done within 1~2 minutes for each step.</p>
<pre><code>subset15_chr21_pbmc_1k_v3_S1_L001_R1_001.fastq.gz
subset15_chr21_pbmc_1k_v3_S1_L001_R2_001.fastq.gz
subset15_chr21_pbmc_1k_v3_S1_L002_R1_001.fastq.gz
subset15_chr21_pbmc_1k_v3_S1_L002_R2_001.fastq.gz</code></pre>
<p>For the mapping, we need a “whitelist” of known cell barcodes. Here we used the 15 barcodes that we already subsetted. The original 737,000 barcodes can be freely extracted from the <a href="https://kb.10xgenomics.com/hc/en-us/articles/115004506263-What-is-a-barcode-whitelist">Cell Ranger pipeline</a>).</p>
<pre><code>subset15_demo_barcode.txt </code></pre>
<p>The barcodes in the R1 FASTQ data are checked against these known cell barcodes in order to assign a specific read to a specific known cell. The barcodes are designed in such a manner that there is virtually no chance that they will align to a place in the reference genome.</p>
<p>In this tutorial we will be using hg19 (GRCh37) version of the human genome, and will therefore also need to use a hg19 GTF fileto annotate our reads (on chromosome 21 only).</p>
<pre><code>chr21.fa
Homo_sapiens.GRCh37.75.21.gtf</code></pre>
<p>10x Genomics has its own processing pipeline, <code>Cell Ranger</code> to process the scRNA-seq outputs it produces, but this process requires much configuration to run and is significantly slower than other mappers.</p>
</div>
<div id="rcwl-version-of-scrna-seq-preprocessing-steps" class="section level3">
<h3 class="hasAnchor">
<a href="#rcwl-version-of-scrna-seq-preprocessing-steps" class="anchor"></a><code>Rcwl</code> version of scRNA-seq preprocessing steps</h3>
<p>In this tutorial, we are trying to reproduce the <a href="https://galaxyproject.github.io/training-material/topics/transcriptomics/tutorials/scrna-preprocessing-tenx/tutorial.html#producing-a-count-matrix-from-fastq">GALAXY pipeline</a> for scRNA-seq data pre-processing. We will use <code>STARsolo</code> to produce a count matrix from FASTQ, and <code>DropletUtils</code> to produce a high quality count matrix with feature/cell annotation files saved in an R object of <code>SingleCellExperiment</code>. Before these 2 steps, we have also added a one-time indexing step in our workshop.</p>
<div id="loading-the-tools" class="section level4">
<h4 class="hasAnchor">
<a href="#loading-the-tools" class="anchor"></a>Loading the tools</h4>
<p>With <code>Rcwl</code>, we can easily write cwl tools and pipelines programmatically in <em>R</em> (Will cover more of this topic later for developers). While we already have the <em>R</em> version of many command line tools in <code>RcwlPipelines</code>, all we need to do is: 1) Search and load the tools/pipelines, 2) assign values for each of the defined parameters, 3) execute the command, and 4) get results saved in user-specified directory, all within <em>R</em>.</p>
<p>Here we show the usage of 3 core functions: <code>cwlUpdate</code>, <code>cwlSearch</code> and <code>cwlInstall</code>.</p>
<ul>
<li><code>cwlUpdate</code></li>
</ul>
<p>The <code>cwlUpdate</code> syncs the current Rcwl recipes and return a <code>BiocFileCache</code> object which contains the most updated Rcwl recipes.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Rcwl</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">RcwlPipelines</span>)</pre></body></html></div>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">atls</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlUpdate.html">cwlUpdate</a></span>()  <span class="co">## sync the tools/pipelines</span>
<span class="no">atls</span></pre></body></html></div>
<pre><code>## class: BiocFileCache
## bfccache: /github/home/.cache/Rcwl
## bfccount: 129
## For more information see: bfcinfo() or bfcquery()</code></pre>
<p>the <code>bfcinfo()</code> function returns a BiocFileCache tibble containing all information about each available tool or pipeline. Currently we have integrated 103 command line tools and 26 pipelines.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu">bfcinfo</span>(<span class="no">atls</span>)</pre></body></html></div>
<pre><code>## # A tibble: 129 x 13
##    rid   rname create_time access_time rpath rtype fpath last_modified_t… etag 
##    &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;            &lt;dbl&gt; &lt;chr&gt;
##  1 BFC1  pl_a… 2020-07-23… 2020-07-23… /git… local /git…               NA &lt;NA&gt; 
##  2 BFC2  pl_A… 2020-07-23… 2020-07-23… /git… local /git…               NA &lt;NA&gt; 
##  3 BFC3  pl_B… 2020-07-23… 2020-07-23… /git… local /git…               NA &lt;NA&gt; 
##  4 BFC4  pl_b… 2020-07-23… 2020-07-23… /git… local /git…               NA &lt;NA&gt; 
##  5 BFC5  pl_b… 2020-07-23… 2020-07-23… /git… local /git…               NA &lt;NA&gt; 
##  6 BFC6  pl_b… 2020-07-23… 2020-07-23… /git… local /git…               NA &lt;NA&gt; 
##  7 BFC7  pl_C… 2020-07-23… 2020-07-23… /git… local /git…               NA &lt;NA&gt; 
##  8 BFC8  pl_G… 2020-07-23… 2020-07-23… /git… local /git…               NA &lt;NA&gt; 
##  9 BFC9  pl_G… 2020-07-23… 2020-07-23… /git… local /git…               NA &lt;NA&gt; 
## 10 BFC10 pl_h… 2020-07-23… 2020-07-23… /git… local /git…               NA &lt;NA&gt; 
## # … with 119 more rows, and 4 more variables: expires &lt;dbl&gt;, Type &lt;chr&gt;,
## #   Command &lt;chr&gt;, Container &lt;chr&gt;</code></pre>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu">bfcinfo</span>(<span class="no">atls</span>)$<span class="no">Type</span>)</pre></body></html></div>
<pre><code>## 
## pipeline     tool 
##       26      103</code></pre>
<ul>
<li><code>cwlSearch</code></li>
</ul>
<p>We can use keywords to search for any specific tools/pipelines of interest, which internally search the keywords against the BiocFileCache tibble in columns of “rname”, “rpath”, “fpath”, “Command” and “Containers”. Multiple keywords are allowed.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">tls</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlSearch.html">cwlSearch</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"STAR"</span>, <span class="st">"index"</span>))
<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">tls</span>)</pre></body></html></div>
<pre><code>##      rid        rname         create_time         access_time
## 1  BFC18 pl_rnaseq_Sf 2020-07-23 19:41:43 2020-07-23 19:41:43
## 2 BFC110 tl_STARindex 2020-07-23 19:42:04 2020-07-23 19:42:04
##                                                             rpath rtype
## 1 /github/home/.cache/Rcwl/RcwlRecipes-master/Rcwl/pl_rnaseq_Sf.R local
## 2 /github/home/.cache/Rcwl/RcwlRecipes-master/Rcwl/tl_STARindex.R local
##                                                             fpath
## 1 /github/home/.cache/Rcwl/RcwlRecipes-master/Rcwl/pl_rnaseq_Sf.R
## 2 /github/home/.cache/Rcwl/RcwlRecipes-master/Rcwl/tl_STARindex.R
##   last_modified_time etag expires     Type
## 1                 NA &lt;NA&gt;      NA pipeline
## 2                 NA &lt;NA&gt;      NA     tool
##                                                                                                                                                                                                                                          Command
## 1 p5&lt;-InputParam(id="in_runThreadN",type="int",default=1L) + fastqc + STAR + runThreadN="in_runThreadN")) + sortBam + samtools_index + samtools_flagstat + featureCounts + RSeQC + gtfToGenePred + genePredToBed + read_distribution + gCoverage
## 2                                                                                                                                                                                                                                           STAR
##                              Container
## 1                                 &lt;NA&gt;
## 2 quay.io/biocontainers/star:2.7.5a--0</code></pre>
<ul>
<li><code>cwlInstall</code></li>
</ul>
<p>The last core function is <code>cwlInstall</code> which sources the Rcwl script for the interested tool/pipeline into the <em>R</em> working environment.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlInstall.html">cwlInstall</a></span>(<span class="no">tls</span>$<span class="no">rname</span>[<span class="fl">2</span>])  <span class="co">## "tl_STARindex"</span>
<span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlInstall.html">cwlInstall</a></span>(<span class="no">tls</span>$<span class="no">fpath</span>[<span class="fl">2</span>])  <span class="co">## equivalent to the above. </span>
<span class="no">STARindex</span></pre></body></html></div>
<pre><code>## class: cwlParam 
##  cwlClass: CommandLineTool 
##  cwlVersion: v1.0 
##  baseCommand: STAR 
## requirements:
## - class: DockerRequirement
##   dockerPull: quay.io/biocontainers/star:2.7.5a--0
## arguments: --runMode genomeGenerate 
## inputs:
##   genomeDir (string): --genomeDir STARindex
##   genomeFastaFiles (File): --genomeFastaFiles 
##   sjdbGTFfile (File): --sjdbGTFfile 
##   runThreadN (int): --runThreadN 4
## outputs:
## outIndex:
##   type: Directory
##   outputBinding:
##     glob: $(inputs.genomeDir)</code></pre>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlParam-methods.html">cwlVersion</a></span>(<span class="no">STARindex</span>)</pre></body></html></div>
<pre><code>## [1] "v1.0"</code></pre>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/InputParamList.html">inputs</a></span>(<span class="no">STARindex</span>)</pre></body></html></div>
<pre><code>## inputs:
##   genomeDir (string): --genomeDir STARindex
##   genomeFastaFiles (File): --genomeFastaFiles 
##   sjdbGTFfile (File): --sjdbGTFfile 
##   runThreadN (int): --runThreadN 4</code></pre>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/OutputParamList.html">outputs</a></span>(<span class="no">STARindex</span>)</pre></body></html></div>
<pre><code>## outputs:
## outIndex:
##   type: Directory
##   outputBinding:
##     glob: $(inputs.genomeDir)</code></pre>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlParam-methods.html">requirements</a></span>(<span class="no">STARindex</span>)</pre></body></html></div>
<pre><code>## [[1]]
## [[1]]$class
## [1] "DockerRequirement"
## 
## [[1]]$dockerPull
## [1] "quay.io/biocontainers/star:2.7.5a--0"</code></pre>
<p>Now the Rcwl version of <code>STARindex</code> is successfully loaded into <em>R</em> and ready to use.</p>
</div>
<div id="indexing" class="section level4">
<h4 class="hasAnchor">
<a href="#indexing" class="anchor"></a>Indexing</h4>
<p>Before the alignment and QC in the aboved mentioned GALAXY pipeline, we will add an indexing step using the tool <code>STARindex</code>. The command line for indexing looks like this:</p>
<pre><code>$ STAR --runMode genomeGenerate --runThreadN 4 --genomeDir STARindex 
--genomeFastaFiles chr21.fa --sjdbGTFfile Homo_sapiens.GRCh37.75.21.gtf</code></pre>
<p>Using the <code>STARindex</code>, we can equivalently do the same indexing within <em>R</em> which was internally passed as a cwl script. What we need to do is assign values for the input parameters, and execute the cwl script using <code>runCWL</code> function.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/InputParamList.html">inputs</a></span>(<span class="no">STARindex</span>)
<span class="no">STARindex</span>$<span class="no">genomeFastaFiles</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="st">"chr21.fa"</span>)
<span class="no">STARindex</span>$<span class="no">sjdbGTFfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="st">"Homo_sapiens.GRCh37.75.21.gtf"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span>(
    <span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/runCWL.html">runCWL</a></span>(<span class="kw">cwl</span> <span class="kw">=</span> <span class="no">STARindex</span>, <span class="kw">outdir</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outpath</span>, <span class="st">"STARindex_output"</span>),
           <span class="kw">docker</span> <span class="kw">=</span> <span class="st">"singularity"</span>, <span class="kw">showLog</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
    ) <span class="co">## elapsed: 74.795S</span>
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outpath</span>, <span class="st">"STARindex_output"</span>), <span class="kw">recursive</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p>Now that the output files are generated in the folder of “STARindex_output” under <code>outpath</code>. These results are ready to pass to the next tool for single cell read alignment.</p>
</div>
<div id="alignment" class="section level4">
<h4 class="hasAnchor">
<a href="#alignment" class="anchor"></a>Alignment</h4>
<p>To align single cell reads using <code>STAR</code>, it can be done directly in command line:</p>
<pre><code>$ STAR --soloType CB_UMI_Simple --genomeDir STARindex_output/STARindex --soloCBwhitelist subset15_demo_barcode.txt --soloUMIlen 12 --readFilesIn subset15_chr21_pbmc_1k_v3_S1_L001_R2_001.fastq.gz,subset15_chr21_pbmc_1k_v3_S1_L002_R2_001.fastq.gz subset15_chr21_pbmc_1k_v3_S1_L001_R1_001.fastq.gz,subset15_chr21_pbmc_1k_v3_S1_L002_R1_001.fastq.gz --soloUMIfiltering MultiGeneUMI --soloCBmatchWLtype 1MM_multi_pseudocounts --readFilesCommand gzcat --runThreadN 1</code></pre>
<p>Now we can follow the previous example to load our Rcwl version of <code>STARsolo</code> and run the command in <em>R</em>.</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlInstall.html">cwlInstall</a></span>(<span class="st">"tl_STARsolo"</span>)  <span class="co">## "tl_STARsolo"</span>
<span class="no">STARsolo</span></pre></body></html></div>
<pre><code>## class: cwlParam 
##  cwlClass: CommandLineTool 
##  cwlVersion: v1.0 
##  baseCommand: STAR 
## requirements:
## - class: DockerRequirement
##   dockerPull: quay.io/biocontainers/star:2.7.5a--0
## arguments: --readFilesCommand zcat --soloUMIfiltering MultiGeneUMI --soloCBmatchWLtype 1MM_multi_pseudocounts 
## inputs:
##   readFilesIn_cdna (File[]): --readFilesIn 
##   readFilesIn_cb (File[]):  
##   genomeDir (Directory): --genomeDir 
##   whiteList (File): --soloCBwhitelist 
##   soloType (string): --soloType CB_UMI_Simple
##   soloUMIlen (string): --soloUMIlen 12
##   runThreadN (int): --runThreadN 1
## outputs:
## outAlign:
##   type: File
##   outputBinding:
##     glob: '*.sam'
## outLog:
##   type: File[]
##   outputBinding:
##     glob: Log*
## SJ:
##   type: File
##   outputBinding:
##     glob: SJ.out.tab
## Solo:
##   type: Directory
##   outputBinding:
##     glob: Solo.out</code></pre>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/InputParamList.html">inputs</a></span>(<span class="no">STARsolo</span>)</pre></body></html></div>
<pre><code>## inputs:
##   readFilesIn_cdna (File[]): --readFilesIn 
##   readFilesIn_cb (File[]):  
##   genomeDir (Directory): --genomeDir 
##   whiteList (File): --soloCBwhitelist 
##   soloType (string): --soloType CB_UMI_Simple
##   soloUMIlen (string): --soloUMIlen 12
##   runThreadN (int): --runThreadN 1</code></pre>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">cdna.fastq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="no">path</span>, <span class="kw">pattern</span> <span class="kw">=</span> <span class="st">"_R2_"</span>))
<span class="no">cb.fastq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="no">path</span>, <span class="kw">pattern</span> <span class="kw">=</span> <span class="st">"_R1_"</span>))
<span class="no">cblist</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="st">"subset15_demo_barcode.txt"</span>)
<span class="no">genomeDir</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="st">"STARindex_output/STARindex"</span>)

<span class="no">STARsolo</span>$<span class="no">readFilesIn_cdna</span> <span class="kw">&lt;-</span> <span class="no">cdna.fastq</span>
<span class="no">STARsolo</span>$<span class="no">readFilesIn_cb</span> <span class="kw">&lt;-</span> <span class="no">cb.fastq</span>
<span class="no">STARsolo</span>$<span class="no">whiteList</span> <span class="kw">&lt;-</span> <span class="no">cblist</span>
<span class="no">STARsolo</span>$<span class="no">genomeDir</span> <span class="kw">&lt;-</span> <span class="no">genomeDir</span></pre></body></html></div>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/runCWL.html">runCWL</a></span>(<span class="no">STARsolo</span>, <span class="kw">outdir</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outpath</span>, <span class="st">"STARsolo_output"</span>), ,
       <span class="kw">docker</span> <span class="kw">=</span> <span class="st">"singularity"</span>, <span class="kw">showLog</span> <span class="kw">=</span> <span class="fl">TRUE</span>)  <span class="co">## approximately 2 mins</span>
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outpath</span>, <span class="st">"STARsolo_output"</span>), <span class="kw">recursive</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p>Now we have all files generated in the “STARsolo_output” folder under <code>outpath</code>, which can be passed into the next tool.</p>
</div>
<div id="dropletutils" class="section level4">
<h4 class="hasAnchor">
<a href="#dropletutils" class="anchor"></a><code>DropletUtils</code>
</h4>
<p>To get a high quality count matrix we must apply the <code>DropletUtils</code> tool, which will produce a filtered dataset that is more representative of the Cell Ranger pipeline.</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlInstall.html">cwlInstall</a></span>(<span class="st">"tl_DropletUtils"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/InputParamList.html">inputs</a></span>(<span class="no">DropletUtils</span>)</pre></body></html></div>
<pre><code>## inputs:
##   dirname (Directory): dir.name= 
##   lower (int): lower= 100
##   df (int): df= 20</code></pre>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="no">DropletUtils</span>$<span class="no">dirname</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outpath</span>, <span class="st">"STARsolo_output/Solo.out"</span>)
<span class="no">DropletUtils</span>$<span class="no">lower</span> <span class="kw">&lt;-</span> <span class="fl">100</span>
<span class="no">DropletUtils</span>$<span class="no">df</span> <span class="kw">&lt;-</span> <span class="fl">5</span></pre></body></html></div>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/runCWL.html">runCWL</a></span>(<span class="no">DropletUtils</span>, <span class="kw">outdir</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outpath</span>, <span class="st">"dropletUtils_output"</span>), <span class="kw">showLog</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outpath</span>, <span class="st">"dropletUtils_output"</span>))</pre></body></html></div>
<p>Now that we get 2 output files:</p>
<ul>
<li>The pdf file with 2 diagnostic figures: Barcode ranks, and empty droplets. Details for each diagnostic figures please refer to the <a href="https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html">DropletUtils vignette</a><br>
</li>
<li>The <code>SingleCellExperiment</code> object which has filtered out unqualified cells.</li>
</ul>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="co">#&gt; Biobase::openPDF(file.path(path, "dropletUtils_output/diagnostics.pdf"))</span>
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outpath</span>, <span class="st">"dropletUtils_output/sce_filtered.rds"</span>))
<span class="no">sce</span></pre></body></html></div>
</div>
</div>
<div id="scrna-seq-preprocessing-pipeline" class="section level3">
<h3 class="hasAnchor">
<a href="#scrna-seq-preprocessing-pipeline" class="anchor"></a>scRNA-seq preprocessing pipeline</h3>
<p>Alternatively and more easily, we can also run the pipeline that was available in <code>RcwlPipelines</code> which has integrated the <code>STARsolo</code> and <code>DropletUtils</code> for a streamlined preprocessing analysis within <em>R</em>.</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlInstall.html">cwlInstall</a></span>(<span class="st">"pl_STARsoloDropletUtils"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/InputParamList.html">inputs</a></span>(<span class="no">STARsoloDropletUtils</span>)
<span class="no">STARsoloDropletUtils</span>$<span class="no">fastq_cdna</span> <span class="kw">&lt;-</span> <span class="no">cdna.fastq</span>
<span class="no">STARsoloDropletUtils</span>$<span class="no">fastq_cb</span> <span class="kw">&lt;-</span> <span class="no">cb.fastq</span>
<span class="no">STARsoloDropletUtils</span>$<span class="no">genomeDir</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outpath</span>, <span class="st">"STARindex_output/STARindex"</span>)
<span class="no">STARsoloDropletUtils</span>$<span class="no">whiteList</span> <span class="kw">&lt;-</span> <span class="no">cblist</span>
<span class="no">STARsoloDropletUtils</span>$<span class="no">runThreadN</span> <span class="kw">&lt;-</span> <span class="fl">2</span></pre></body></html></div>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/runCWL.html">runCWL</a></span>(<span class="no">STARsoloDropletUtils</span>, <span class="kw">outdir</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outpath</span>, <span class="st">"scpipeline_output"</span>),
       <span class="kw">showLog</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">outpath</span>, <span class="st">"scpipeline_output"</span>), <span class="kw">recursive</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/OutputParamList.html">outputs</a></span>(<span class="no">STARsoloDropletUtils</span>)</pre></body></html></div>
</div>
<div id="todo" class="section level3">
<h3 class="hasAnchor">
<a href="#todo" class="anchor"></a>todo:</h3>
<ul>
<li>explain the output globs.</li>
<li>Explain the cwl files. (config, values)</li>
<li>test docker.</li>
</ul>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Qian Liu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
